src/filters/WorldFilter.cpp

```diff
@@ -0,0 +1,138 @@
+//
+// Created by kjhertenberg on 13-5-19.
+//
+
+#include "roboteam_proto/messages_robocup_ssl_detection.pb.h"
+#include <filters/WorldFilter.h>
+#include <memory>
+
+namespace world {
+
+    WorldFilter::WorldFilter() {
+        blueBots.clear();
+        yellowBots.clear();
+        balls.clear();
+    }
+
+// if we get a new frame we update our observations
+    void WorldFilter::addFrame(const proto::SSL_DetectionFrame &msg) {
+        const double filterGrabDistance = 0.5;
+        double timeCapture = msg.t_capture();
+        uint cameraID = msg.camera_id();
+        handleRobots(yellowBots,msg.robots_yellow(), filterGrabDistance, timeCapture, cameraID);
+        handleRobots(blueBots,msg.robots_blue(), filterGrabDistance, timeCapture, cameraID);
+        handleBall(msg.balls(), filterGrabDistance, timeCapture, cameraID);
+    }
+    void
+    WorldFilter::handleBall(const google::protobuf::RepeatedPtrField<proto::SSL_DetectionBall> &observations, const double filterGrabDistance, double timeCapture,
+                            uint cameraID) {
+        for (const proto::SSL_DetectionBall &detBall :observations) {
+            bool addedBall = false;
+            for (const auto &filter : balls) {
+                if (filter->distanceTo(detBall.x(), detBall.y()) < filterGrabDistance) {
+                    filter->addObservation(detBall, timeCapture, cameraID);
+                    addedBall = true;
+                }
+            }
+            if (!addedBall) {
+                balls.push_back(std::make_unique<BallFilter>(detBall, timeCapture, cameraID));
+            }
+        }
+    }
+    void
+    WorldFilter::handleRobots(robotMap &robots,
+                              const google::protobuf::RepeatedPtrField<proto::SSL_DetectionRobot> &observations,
+                              double filterGrabDistance, double timeCapture, uint cameraID) {
+        for (const proto::SSL_DetectionRobot &robot : observations) {
+            bool addedBot = false;
+            for (const auto &filter : robots[robot.robot_id()]) {
+                if (filter->distanceTo(robot.x(), robot.y()) < filterGrabDistance) {
+                    filter->addObservation(robot, timeCapture, cameraID);
+                    addedBot = true;
+                }
+            }
+            if (!addedBot) {
+                // We create a new filter if no filter close to the robot exists
```

Comment:
I think "creating a filter" can be a bit confusing. You are creating a filter that tracks that observation if it is not already within a certain distance of another state being tracked by a filter, if I understand what you are doing correctly.

Reply:
I mean the filter is literally constructed so in that sense it makes sense. The way I view and the way it happens in this architecture is that the robotfilters are not responsible for selecting the right observations but the worldFilter is. In that sense 'creating' a filter makes a lot more sense, if you view WorldFilter as an entity that manages other filters and which detections to send to which 'lower-level' filter.